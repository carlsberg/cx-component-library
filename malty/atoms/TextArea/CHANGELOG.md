# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2024-05-31)

### Bug Fixes

- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- **TextArea:** [DSM-563] Update counter when receive initial value ([#574](https://github.com/CarlsbergGBS/cx-component-library/issues/574)) ([6a5d360](https://github.com/CarlsbergGBS/cx-component-library/commit/6a5d3601a297fa3f26de7bd7cd5a20b048c48d77))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-235:** added readOnly on inputs ([#181](https://github.com/CarlsbergGBS/cx-component-library/issues/181)) ([36c5ef7](https://github.com/CarlsbergGBS/cx-component-library/commit/36c5ef7cb1225bf5a150918e01e10fd253377110))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-368:** Required prop & label component ([#384](https://github.com/CarlsbergGBS/cx-component-library/issues/384)) ([983bd78](https://github.com/CarlsbergGBS/cx-component-library/commit/983bd78c8f526a8faa2df601bf918b40219820f2))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** Remove `react/jsx-props-no-spreading` comments ([#558](https://github.com/CarlsbergGBS/cx-component-library/issues/558)) ([823444c](https://github.com/CarlsbergGBS/cx-component-library/commit/823444ce0800bc7f49b584de59d247eb8497f559))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- **MS-108:** added test id to text area ([#374](https://github.com/CarlsbergGBS/cx-component-library/issues/374)) ([b06baa1](https://github.com/CarlsbergGBS/cx-component-library/commit/b06baa132edc1dd2804ab54fa6ceda236ae80431))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Several component updates ([#185](https://github.com/CarlsbergGBS/cx-component-library/issues/185)) ([633c9f0](https://github.com/CarlsbergGBS/cx-component-library/commit/633c9f0a4432e9e3d08ff05156453660929c19a2))
- **textarea:** [DSM-797] add tests ([#766](https://github.com/CarlsbergGBS/cx-component-library/issues/766)) ([939b83a](https://github.com/CarlsbergGBS/cx-component-library/commit/939b83aa215d33d09d83d07349c1fddda9d44672))
- **textarea:** [DSM-972] uniformize styles ([#873](https://github.com/CarlsbergGBS/cx-component-library/issues/873)) ([6be0f5c](https://github.com/CarlsbergGBS/cx-component-library/commit/6be0f5c35d89a038f1e203c14190e347105bb8eb))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
