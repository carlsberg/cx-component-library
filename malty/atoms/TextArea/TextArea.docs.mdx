---
labels: ['react', 'typescript', 'textarea']
description: 'A textarea component.'
---
