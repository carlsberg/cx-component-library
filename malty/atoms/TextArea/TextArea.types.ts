export interface TextAreaProps extends React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  value: string;
  label?: string;
  placeholder?: string;
  resize?: boolean;
  disabled?: boolean;
  onValueChange: (value: string) => void;
  error?: string;
}
