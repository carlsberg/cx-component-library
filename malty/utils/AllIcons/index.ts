export { allIcons, allIconsStoryOptions } from './AllIconsMap';
