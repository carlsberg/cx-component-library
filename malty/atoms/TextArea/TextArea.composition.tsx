import React from 'react';
import { TextArea } from './TextArea';

export const BasicTextArea = () => (
  <TextArea value="TextArea value" label="TextArea label" onValueChange={() => null} />
);
