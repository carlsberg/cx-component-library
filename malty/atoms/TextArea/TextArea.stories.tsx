import { Meta, Story } from '@storybook/react';
import React, { useState } from 'react';
import { TextArea as TextAreaComponent } from './TextArea';
import { TextAreaProps } from './TextArea.types';

export default {
  title: 'Forms/Text Area',
  component: TextAreaComponent,
  parameters: {
    importObject: 'TextArea',
    importPath: '@carlsberggroup/malty.atoms.text-area'
  },
  argTypes: {
    label: {
      description: 'Label for the input, goes above.',
      control: 'text'
    },
    placeholder: {
      description: 'Placeholder text to go inside the input field, when empty.',
      control: 'text'
    },
    resize: {
      control: 'boolean',
      description: 'Textarea state, resizable textarea'
    },
    disabled: {
      control: 'boolean',
      description: 'Textarea state, disabled'
    },
    readOnly: {
      control: 'boolean',
      description: 'Textarea state, readOnly'
    },
    error: {
      description: 'Error message to be displayed when error is present.',
      control: 'text'
    },
    hint: {
      description: 'helper message to be displayed',
      control: 'text'
    },
    maxLength: {
      description: 'Maxium characters of textarea ',
      control: 'number'
    },

    value: {
      description: 'Default value of textarea',
      control: 'text'
    },
    onValueChange: {
      description: 'Function to be executed when textarea state changes'
    },
    dataTestId: {
      control: 'text',
      description: 'select data-testid'
    },
    required: {
      control: 'boolean',
      description: 'Makes the textarea input required to fill'
    }
  }
} as Meta;

const Template: Story<TextAreaProps> = ({ value, onValueChange, ...args }) => {
  const [stateValue, setStateValue] = useState(value);
  return (
    <TextAreaComponent value={stateValue} onValueChange={(newValue: string) => setStateValue(newValue)} {...args} />
  );
};

export const TextArea = Template.bind({});

const params = new URLSearchParams(window.location.search);
const error = params.get('error');
const resizable = params.get('resizable');

TextArea.args = {
  label: 'Label',
  maxLength: 20,
  resize: !!resizable,
  placeholder: 'Placeholder',
  disabled: false,
  error: error ? 'Error text' : '',
  value: '',
  hint: 'hint text',
  readOnly: false,
  dataTestId: 'Textarea',
  required: false
};
