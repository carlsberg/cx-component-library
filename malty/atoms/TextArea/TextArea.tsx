import { globalTheme as defaultTheme, TypographyProvider } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext, useMemo, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { v4 as uuid } from 'uuid';
import {
  StyledError,
  StyledHint,
  StyledLabel,
  StyledtextArea,
  StyledTextAreaCharacterCounter,
  StyledTextareaContainer,
  StyledTextAreaWrapper
} from './TextArea.styled';
import { TextAreaProps } from './TextArea.types';

export const TextArea = ({
  label,
  placeholder,
  resize = false,
  disabled = false,
  value,
  onValueChange,
  error,
  maxLength,
  hint,
  readOnly = false,
  required,
  ...props
}: TextAreaProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;
  const id = useMemo(() => uuid(), []);
  const [textAreaCount, setTextAreaCount] = useState(maxLength || 0);

  const handleCarachterCounter = (e: React.FormEvent<HTMLTextAreaElement>) => {
    if (maxLength) {
      setTextAreaCount(maxLength - e.currentTarget.value.length);
    } else {
      setTextAreaCount(e.currentTarget.value.length);
    }
    onValueChange(e.currentTarget.value as string);
  };

  return (
    <TypographyProvider>
      <StyledTextareaContainer theme={theme}>
        {label && (
          <StyledLabel required={required} disabled={disabled} htmlFor={id} theme={theme}>
            {label}
          </StyledLabel>
        )}
        <StyledTextAreaWrapper readOnly={readOnly} disabled={disabled} isError={!!error} resize={resize} theme={theme}>
          <StyledtextArea
            name={id}
            id={id}
            value={value}
            placeholder={placeholder}
            onChange={handleCarachterCounter}
            theme={theme}
            disabled={disabled}
            readOnly={readOnly}
            maxLength={maxLength}
            required={required}
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...props}
          />
          <StyledTextAreaCharacterCounter disabled={disabled} theme={theme} data-testid="TextAreaCounter">
            {textAreaCount}
          </StyledTextAreaCharacterCounter>
        </StyledTextAreaWrapper>
        {error && <StyledError theme={theme}>{error}</StyledError>}
        {hint && !error && (
          <StyledHint disabled={disabled} theme={theme}>
            {hint}
          </StyledHint>
        )}
      </StyledTextareaContainer>
    </TypographyProvider>
  );
};
