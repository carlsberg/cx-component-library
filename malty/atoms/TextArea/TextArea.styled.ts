import { space, SpaceProps } from '@carlsberggroup/malty.utils.space';
import styled, { css } from 'styled-components';

export const StyledTextareaContainer = styled.div<SpaceProps>`
  display: flex;
  flex: 1 1 auto;
  flex-direction: column;

  ${space}
`;

export const StyledError = styled.label`
  color: ${({ theme }) => theme.colors.colours.system.fail.value};
  font-family: ${({ theme }) => theme.typography.desktop.text.tiny_default['font-family'].value};
  font-size: ${({ theme }) => theme.typography.desktop.text.tiny_default['font-size'].value};
  font-weight: bold;
  line-height: ${({ theme }) => theme.typography.desktop.text.tiny_default['line-height'].value};
  letter-spacing: 0;
  margin-top: ${({ theme }) => theme.sizes['4xs'].value};
`;
export const StyledHint = styled.label<{
  disabled?: boolean;
}>`
  color: ${({ theme }) => theme.colors.colours.support[60].value};
  font-family: ${({ theme }) => theme.typography.desktop.text.tiny_default['font-family'].value};
  font-size: ${({ theme }) => theme.typography.desktop.text.tiny_default['font-size'].value};
  font-weight: bold;
  line-height: ${({ theme }) => theme.typography.desktop.text.tiny_default['line-height'].value};
  letter-spacing: 0;
  margin-top: ${({ theme }) => theme.sizes['4xs'].value};
  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.colors.colours.system['disable-light-theme'].value};
    `}
`;

export const StyledtextArea = styled.textarea<{
  disabled?: boolean;
  readOnly?: boolean;
}>`
  width: 100%;
  height: calc(100% - 22px);
  box-sizing: border-box;
  font-weight: normal;
  font-family: ${({ theme }) => theme.typography.desktop.text['medium-small_default']['font-family'].value};
  font-size: ${({ theme }) => theme.typography.desktop.text['medium-small_default']['font-size'].value};
  transition: 0.25s ease-in-out;
  transition-property: border-color, color;
  color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};
  ${({ theme }) =>
    css`
      padding: ${theme.sizes.xs.value} ${theme.sizes.xs.value};
    `}
  resize: none;
  border: 0;
  &:hover,
  &:focus {
    outline: none;
  }

  ::placeholder {
    opacity: 0.8;
    color: ${({ theme }) => theme.colors.colours.information.indirect.value};
    ${({ disabled, theme }) =>
      disabled &&
      css`
        color: ${theme.colors.colours.system['disable-light-theme'].value};
      `}
  }
  ${({ disabled, theme }) =>
    disabled &&
    css`
      background-color: ${theme.colors.colours.default.transparent.value};
      color: ${theme.colors.colours.system['disable-light-theme'].value};
    `}
  ${({ readOnly, theme }) =>
    readOnly &&
    css`
      background-color: ${theme.colors.colours.support[20].value};
      color: ${theme.colors.colours.support[80].value};
    `}
`;

export const StyledTextAreaWrapper = styled.div<{
  resize?: boolean;
  disabled?: boolean;
  readOnly?: boolean;
  isError?: boolean;
}>`
  min-height: 96px;
  height: 100px;
  position: relative;
  display: flex;
  flex: 1 1 auto;
  flex-direction: column;
  border: 1px solid
    ${({ theme, isError }) =>
      isError ? theme.colors.colours.system.fail.value : theme.colors.colours.support[40].value};
  overflow: hidden;
  background-color: ${({ theme }) => theme.colors.colours.default.white.value};
  &:hover,
  &:focus {
    outline: none;
  }
  &:hover {
    border-color: ${({ theme }) => theme.colors.colours.information.indirect.value};
  }
  &:focus {
    border-color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};
    color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};
  }
  ${({ resize }) =>
    resize
      ? css`
          resize: vertical;
        `
      : css`
          resize: none;
        `}
  ${({ theme, disabled }) =>
    disabled &&
    css`
      border-color: ${theme.colors.colours.system['disable-light-theme'].value};
      color: ${theme.colors.colours.system['disable-light-theme'].value};
      &:hover,
      &:focus {
        outline: none;
        border-color: inherit;
      }
    `}
  ${({ theme, readOnly }) =>
    readOnly &&
    css`
      border-color: ${theme.colors.colours.support[40].value};
      color: ${theme.colors.colours.support[80].value};
      background-color: ${theme.colors.colours.support[20].value};
      &:hover,
      &:focus {
        outline: none;
        border-color: ${theme.colors.colours.support[40].value};
      }
    `}
`;
export const StyledTextAreaCharacterCounterContainer = styled.div`
  position: relative;
  display: flex;
  flex: 1 1 auto;
`;
export const StyledTextAreaCharacterCounter = styled.div<{
  disabled?: boolean;
}>`
  margin-top: 5px;
  position: relative;
  width: fit-content;
  margin-left: 8px;
  margin-bottom: 8px;
  background-color: ${({ theme }) => theme.colors.colours.support[60].value};
  color: ${({ theme }) => theme.colors.colours.default.white.value};
  font-family: ${({ theme }) => theme.typography.desktop.text.tiny_default['font-family'].value};
  font-size: ${({ theme }) => theme.typography.desktop.text.tiny_default['font-size'].value};
  padding: 0 ${({ theme }) => theme.sizes['3xs'].value};
  border-radius: 7px;
  display: flex;
  align-items: center;
  height: 14px;
  font-weight: bold;
  ${({ disabled }) =>
    disabled &&
    css`
      background-color: ${({ theme }) => theme.colors.colours.system['disable-light-theme'].value};
    `}
`;
