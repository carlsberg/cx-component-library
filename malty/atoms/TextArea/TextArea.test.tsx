import { render } from '@carlsberggroup/malty.utils.test';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { TextArea } from './TextArea';

jest.mock('uuid', () => ({ v4: () => '00000000-0000-0000-0000-000000000000' }));

const mockFn = jest.fn();

describe('textarea', () => {
  it('renders elements', () => {
    render(
      <TextArea
        value="Value text"
        label="Label text"
        placeholder="Placeholder text"
        onValueChange={mockFn}
      />
    );
    expect(screen.getByLabelText('Label text')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Placeholder text')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Value text')).toBeInTheDocument();
  });

  it('calls onValueChange when typing', () => {
    const onValueChange = jest.fn();
    const { rerender } = render(
      <TextArea value="Initial value" label="textarea label" onValueChange={onValueChange} />
    );
    const textarea = screen.getByDisplayValue('Initial value');
    fireEvent.input(textarea, { target: { value: 'Test' } });
    expect(onValueChange).toHaveBeenCalledTimes(1);

    rerender(<TextArea value="Test" label="textarea label" onValueChange={onValueChange} />);
    expect(screen.getByDisplayValue('Test')).toBeInTheDocument();
  });
});
