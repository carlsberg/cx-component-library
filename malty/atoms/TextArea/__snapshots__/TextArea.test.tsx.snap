// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textarea matches snapshot 1`] = `
<span
  className="sc-bdvvtL bWTUHG"
>
  <div
    className="sc-gsDKAQ IhRIv"
  >
    <label
      className="sc-dkPtRN LDOZH"
      htmlFor="00000000-0000-0000-0000-000000000000"
    >
      Label text
    </label>
    <div
      className="sc-gKclnd dvYTVD"
    >
      <textarea
        className="sc-jRQBWg exRojm"
        id="00000000-0000-0000-0000-000000000000"
        name="00000000-0000-0000-0000-000000000000"
        onChange={[Function]}
        value="Value"
      />
      <div
        className="sc-furwcr eZoIgo"
        data-testid="TextAreaCounter"
      >
        0
      </div>
    </div>
  </div>
</span>
`;
